test/framework/src/main/java/org/elasticsearch/transport/nio/channel/NioSocketChannel.java

```diff
@@ -0,0 +1,187 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.transport.nio.channel;
+
+import org.elasticsearch.transport.nio.ESSelector;
+import org.elasticsearch.transport.nio.NetworkBytesReference;
+import org.elasticsearch.transport.nio.SocketSelector;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.nio.ByteBuffer;
+import java.nio.channels.ClosedChannelException;
+import java.nio.channels.SocketChannel;
+
+public class NioSocketChannel extends AbstractNioChannel<SocketChannel> {
+
+    private final InetSocketAddress remoteAddress;
+    private final ConnectFuture connectFuture = new ConnectFuture();
+    private WriteContext writeContext;
+    private ReadContext readContext;
+    private SocketSelector socketSelector;
+
+    public NioSocketChannel(String profile, SocketChannel socketChannel) throws IOException {
+        super(profile, socketChannel);
+        this.remoteAddress = (InetSocketAddress) socketChannel.getRemoteAddress();
+    }
+
+    @Override
+    public CloseFuture closeAsync() {
+        // Even if the channel has already been closed we will clear any pending write operations just in case
+        if (state.get() > UNREGISTERED && getSelector().isOnCurrentThread() && writeContext.hasQueuedWriteOps()) {
+            writeContext.clearQueuedWriteOps(new ClosedChannelException());
+        }
+
+        return super.closeAsync();
+    }
+
+    @Override
+    public void closeFromSelector() {
+        // Even if the channel has already been closed we will clear any pending write operations just in case
+        if (state.get() > UNREGISTERED && getSelector().isOnCurrentThread() && writeContext.hasQueuedWriteOps()) {
+            writeContext.clearQueuedWriteOps(new ClosedChannelException());
+        }
+
+        super.closeFromSelector();
+    }
+
+    @Override
+    public SocketSelector getSelector() {
+        return socketSelector;
+    }
+
+    @Override
+    boolean markRegistered(ESSelector selector) {
+        this.socketSelector = (SocketSelector) selector;
+        return super.markRegistered(selector);
+    }
+
+    public int write(NetworkBytesReference reference) throws IOException {
```

Comment:
I think here we can go with an ordinary `BytesReference` and use it's efficient iterator `BytesRefIterator iterator()` the returned `BytesRef` is just a pointer into the underlying `byte[]` that you can wrap in with `ByteBuffer#wrap` and do your accounting what is left for writing in this method. so there is no need to use a `NetworkBytesReference`